kernel/kernel-api/src/main/java/io/delta/kernel/internal/stats/FileSizeHistogram.java

```diff
@@ -0,0 +1,222 @@
+        124 * MB,
+        128 * MB,
+        132 * MB,
+        136 * MB,
+        140 * MB,
+        144 * MB,
+        // 16 MB jumps till 576 MB
+        160 * MB,
+        176 * MB,
+        192 * MB,
+        208 * MB,
+        224 * MB,
+        240 * MB,
+        256 * MB,
+        272 * MB,
+        288 * MB,
+        304 * MB,
+        320 * MB,
+        336 * MB,
+        352 * MB,
+        368 * MB,
+        384 * MB,
+        400 * MB,
+        416 * MB,
+        432 * MB,
+        448 * MB,
+        464 * MB,
+        480 * MB,
+        496 * MB,
+        512 * MB,
+        528 * MB,
+        544 * MB,
+        560 * MB,
+        576 * MB,
+        // 64 MB jumps till 1408 MB
+        640 * MB,
+        704 * MB,
+        768 * MB,
+        832 * MB,
+        896 * MB,
+        960 * MB,
+        1024 * MB,
+        1088 * MB,
+        1152 * MB,
+        1216 * MB,
+        1280 * MB,
+        1344 * MB,
+        1408 * MB,
+        // 128 MB jumps till 2 GB
+        1536 * MB,
+        1664 * MB,
+        1792 * MB,
+        1920 * MB,
+        2048 * MB,
+        // 256 MB jumps till 4 GB
+        2304 * MB,
+        2560 * MB,
+        2816 * MB,
+        3072 * MB,
+        3328 * MB,
+        3584 * MB,
+        3840 * MB,
+        4 * GB,
+        // power of 2 till 256 GB
+        8 * GB,
+        16 * GB,
+        32 * GB,
+        64 * GB,
+        128 * GB,
+        256 * GB);
+  }
+
+  public static FileSizeHistogram init() {
+    return new FileSizeHistogram(
+        DEFAULT_BIN_BOUNDARIES,
+        initializeZeroArray(DEFAULT_BIN_BOUNDARIES.size()),
+        initializeZeroArray(DEFAULT_BIN_BOUNDARIES.size()));
+  }
+
+  /**
+   * Creates a histogram with custom boundaries and initial counts.
+   *
+   * @param sortedBinBoundaries List of boundary values in ascending order
+   * @param fileCounts Array of file counts for each bin
+   * @param totalBytes Array of total bytes for each bin
+   * @throws IllegalArgumentException if arrays have different lengths
+   */
+  public FileSizeHistogram(List<Long> sortedBinBoundaries, Long[] fileCounts, Long[] totalBytes) {
+    this.sortedBinBoundaries =
+        new ArrayList<>(requireNonNull(sortedBinBoundaries, "sortedBinBoundaries cannot be null"));
+    this.fileCounts =
+        Arrays.copyOf(requireNonNull(fileCounts, "fileCounts cannot be null"), fileCounts.length);
+    this.totalBytes =
+        Arrays.copyOf(requireNonNull(totalBytes, "totalBytes cannot be null"), totalBytes.length);
+
+    if (sortedBinBoundaries.size() != fileCounts.length
+        || sortedBinBoundaries.size() != totalBytes.length) {
+      throw new IllegalArgumentException("All arrays must have the same length");
+    }
+  }
+
+  /**
+   * Adds a file size to the histogram, incrementing the appropriate bin's count and total bytes.
+   *
+   * @param fileSize The size of the file in bytes
+   * @throws IllegalArgumentException if fileSize is negative
+   */
+  public void insert(long fileSize) {
```

Comment:
what do we do if filesize doesn't fit into _any_ bin?

Reply:
Added a comment

In the current implementation, all size will find a bin. If too large, it will go to the last one